examples/rust/tcp-ping-pong.rs

```diff
@@ -198,178 +197,174 @@ fn close(libos: &mut LibOS, sockqd: QDesc) {
+            Err(e) => anyhow::bail!("failed to create socket: {:?}", e),
+        };
 
-    // Setup peer.
-    let sockqd: QDesc = match libos.socket(AF_INET, SOCK_STREAM, 0) {
-        Ok(sockqd) => sockqd,
-        Err(e) => anyhow::bail!("failed to create socket: {:?}", e),
-    };
-    match libos.bind(sockqd, local) {
-        Ok(()) => (),
-        Err(e) => {
-            // If error, free socket.
-            close(&mut libos, sockqd);
-            anyhow::bail!("bind failed: {:?}", e)
-        },
-    };
+        return Ok(Self { libos, sockqd });
+    }
 
-    // Mark as a passive one.
-    match libos.listen(sockqd, 16) {
-        Ok(()) => (),
-        Err(e) => {
-            // If error, free socket.
-            close(&mut libos, sockqd);
-            anyhow::bail!("listen failed: {:?}", e)
-        },
-    };
+    pub fn run(&mut self, local: SocketAddrV4, nrounds: usize) -> Result<()> {
+        match self.libos.bind(self.sockqd, local) {
+            Ok(()) => (),
+            Err(e) => anyhow::bail!("bind failed: {:?}", e),
+        };
 
-    // Accept incoming connections.
-    // If error, free socket.
-    let sockqd: QDesc = match accept_and_wait(&mut libos, sockqd) {
-        Ok(sockqd) => sockqd,
-        Err(e) => {
-            close(&mut libos, sockqd);
-            anyhow::bail!("accept failed: {:?}", e)
-        },
-    };
+        // Mark as a passive one.
+        match self.libos.listen(self.sockqd, 16) {
+            Ok(()) => (),
+            Err(e) => anyhow::bail!("listen failed: {:?}", e),
+        };
 
-    // Perform multiple ping-pong rounds.
-    for i in 0..nrounds {
-        let mut fill_char: u8 = (i % (u8::MAX as usize - 1) + 1) as u8;
-
-        // Pop data, and sanity check it.
-        {
-            let mut recvbuf: [u8; BUFFER_SIZE] = [0; BUFFER_SIZE];
-            // If error, free socket.
-            if let Err(e) = pop_and_wait(&mut libos, sockqd, &mut recvbuf) {
-                close(&mut libos, sockqd);
-                anyhow::bail!("pop and wait failed: {:?}", e);
-            }
+        // Accept incoming connections.
+        let accepted_qd: QDesc = match accept_and_wait(&mut self.libos, self.sockqd) {
+            Ok(qd) => qd,
+            Err(e) => anyhow::bail!("accept failed: {:?}", e),
+        };
+
+        // Perform multiple ping-pong rounds.
+        for i in 0..nrounds {
+            let mut fill_char: u8 = (i % (u8::MAX as usize - 1) + 1) as u8;
 
-            for x in &recvbuf {
-                if *x != fill_char {
-                    close(&mut libos, sockqd);
-                    anyhow::bail!("fill check failed: expected={:?} received={:?}", fill_char, *x);
+            // Pop data, and sanity check it.
+            {
+                let mut recvbuf: [u8; BUFFER_SIZE] = [0; BUFFER_SIZE];
+                if let Err(e) = pop_and_wait(&mut self.libos, accepted_qd, &mut recvbuf) {
+                    close(&mut self.libos, accepted_qd);
```

Comment:
You could add `accepted_qd: Option<QDesc>` to the server data structure and then close it in the drop if there is qd. That way we won't have to specifically call close each time here and on line 252.

Reply:
Done.